File: Java/Questao4_JantarFilosofos.java
import java.util.concurrent.Semaphore;

public class Questao4_JantarFilosofos {
    static final int N = 5;
    static Semaphore[] garfos = new Semaphore[N];

    static class Filosofo extends Thread {
        int id;
        Filosofo(int id) { this.id = id; }

        public void run() {
            try {
                while (true) {
                    System.out.println("Filósofo " + id + " pensando...");
                    Thread.sleep(1000);

                    garfos[id].acquire();
                    garfos[(id+1)%N].acquire();

                    System.out.println("Filósofo " + id + " comendo...");
                    Thread.sleep(2000);

                    garfos[id].release();
                    garfos[(id+1)%N].release();

                    System.out.println("Filósofo " + id + " terminou de comer.");
                }
            } catch (InterruptedException e) {}
        }
    }

    public static void main(String[] args) {
        for (int i = 0; i < N; i++)
            garfos[i] = new Semaphore(1);

        for (int i = 0; i < N; i++)
            new Filosofo(i).start();
    }
}

File: C/Questao4_JantarFilosofos.c
#include <stdio.h>
#include <pthread.h>
#include <semaphore.h>
#include <unistd.h>

#define N 5

sem_t forks[N];

void* filosofo(void* arg) {
    int id = *((int*)arg);

    while (1) {
        printf("Filósofo %d pensando...\n", id);
        sleep(1);

        sem_wait(&forks[id]);
        sem_wait(&forks[(id+1)%N]);

        printf("Filósofo %d comendo...\n", id);
        sleep(2);

        sem_post(&forks[id]);
        sem_post(&forks[(id+1)%N]);

        printf("Filósofo %d terminou de comer.\n", id);
    }
    return NULL;
}

int main() {
    pthread_t filosofos[N];
    int id[N];

    for (int i = 0; i < N; i++) {
        sem_init(&forks[i], 0, 1);
        id[i] = i;
        pthread_create(&filosofos[i], NULL, filosofo, &id[i]);
    }

    for (int i = 0; i < N; i++)
        pthread_join(filosofos[i], NULL);

    return 0;
}
